### PR TITLE
SQL: Implement NULLIF(expr1, expr2) function

### PR DESCRIPTION
**Pull request by matriv**
NULLIF returns null if the 2 expressions are equal or the
expr1 otherwise.

Closes: #35818

### CODE REVIEW

File: docs/reference/sql/functions/conditional.asciidoc [hunk 1]
```diff
@@ -155,3 +155,38 @@ include-tagged::{sql-specs}/docs.csv-spec[nvlReturnFirst]
...
+----
+
+*Input*:
+
+<1> 1st expression
+
+<2> 2nd expression
+
+
+*Output*: `NULL` if the 2 expressions are equal, otherwise the 1st expression.
```
Comment (astefan) [inline docs/reference/sql/functions/conditional.asciidoc]: Lowercase `null`? For another function doc page you used the lowercased version of `null`: https://github.com/elastic/elasticsearch/pull/35793/files#diff-cf102e856ee9e39135c9677a52db35e1R108

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NULLIf.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.elasticsearch.xpack.sql.expression.gen.script.ParamsBuilder.paramsBuilder;
+
+/**
+ * Name is `NULLIf` to avoid having it registered as `NULL_IF` instead of `NULLIF`.
+ */
+public class NULLIf extends ConditionalFunction {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NULLIf.java]: Can we do something about these class names? I understand why it is named like this and same goes for `IFNull` added previously, but I believe we should change the way we normalize the names. Simplest/quickest approach: put these exception functions in a Set (or similar) and treat them differently in the `normalize` method from `FunctionRegistry`.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIfProcessor.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    @Override
+    public Object process(Object input) {
+        Object leftValue = leftProcessor.process(input);
+        Object rightValue = rightProcessor.process(input);
+        return apply(leftValue, rightValue);
+    }
+
+    public static Object apply(Object leftValue, Object rightValue) {
+        if (BinaryComparisonProcessor.BinaryComparisonOperation.EQ.apply(leftValue, rightValue) == Boolean.TRUE) {
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/conditional/NullIfProcessor.java]: Maybe `import org.elasticsearch.xpack.sql.expression.predicate.operator.comparison.BinaryComparisonProcessor.BinaryComparisonOperation` to shorten this line and make it more readable?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 4]
```diff
@@ -1172,6 +1174,9 @@ protected Expression rule(Expression e) {
                     return Literal.of(in, null);
                 }
 
+            } else if (e instanceof NULLIf) {
+                return e; // Special rule SimplifyNullIf is applied instead
+
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Remove this empty line.

**Review by astefan — COMMENTED** [review]
Looking good. Left few comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java (hunk 4 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Was on purpose as we have them in all `else if` blocks.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
retest this please

**Comment by matriv** [comment]
retest this please

**Review by costin — APPROVED** [review]
LGTM

**Comment by matriv** [comment]
Backported to `6.x` with 47575567fa66444912f9785e05f85fbc646faa30

**Pull request merged**